### ISSUE TITLE
Populating across Databases using Dynamic References

### ISSUE THREAD

**Issue by mushfau** [Automattic/mongoose#6520]
[](url)
**Do you want to request a *feature* or report a *bug*
**BUG** or a **FEATURE**

**What is the current behavior?**
Populating dynamically referenced cross-database models throws MissingSchemaError

> MissingSchemaError: Schema hasn't been registered for model "Book". 
Use mongoose.model(name, schema)

it works as needed if the referenced models are in the same database

**If the current behavior is a bug, please provide the steps to reproduce.**

the behavior is reproduced in the gist available [here](https://gist.github.com/mushfau/06a36e243c2831183f3bda1b94843731)

**What is the expected behavior?**
populate the the dynamically referenced models even though they are in across databases 

**Please mention your node.js, mongoose and MongoDB version.**
MongoDB version: 3.4.3
Mongoose version: 5.1.2
NodeJS version: 8.11.1

**Comment by vkarpov15** [comment]
We currently don't support this. [Cross-db populate](http://mongoosejs.com/docs/populate.html#cross-db-populate) only works if you pass in the whole mongoose model object to `populate()`. You can't store mongoose models in mongodb, so dynref relies fully on model names, and when you pass a model name to `populate()` mongoose only looks in the current connection for models with that name.

We'll add support for this, shouldn't be too hard. I think the idea will be to add a `connection` option to `populate()` that lets you specify the [connection](http://mongoosejs.com/docs/connections.html) to use for looking up model names.  The code would look like this:

```javascript
// Tell mongoose to look in `hobby_db` for models named 'Movie' and 'User'
await User.find({}).populate({ path: 'hobby', connection: hobby_db });
```

**Comment by vkarpov15** [comment]
Almost fixed but not quite, there's an issue related to #5858 that's still popping up intermittently, looks like there's a subtle race condition somewhere.

**Comment by taxilian** [comment]
FYI, the fix for this issue (specifically commit a33c82eef7bd1768f727f2004f96cd4115157f91) has broken something else; when we populate an array field it no longer correctly gives us an array of populated documents if there is only one document in the array, it just gives us the document. I'm going to try to build a test case showing the issue and submit a separate issue

**Comment by vkarpov15** [comment]
@taxilian I'm guessing this is the issue you fixed with #8125?

**Comment by taxilian** [comment]
Yes it is
